bcda/models/postgres/alr.go

```diff
@@ -0,0 +1,252 @@
+}
+
+func (r *AlrRepository) AddAlr(ctx context.Context, aco string, timestamp time.Time, alrs []models.Alr) error {
+	// Do this in a single transaction using BeginEX from the pgx package
+	tx, err := r.BeginEx(ctx, nil)
+	if err != nil {
+		return err
+	}
+
+	// Update the alr_meta table and get the foreign key
+	updateAlrMeta := sqlFlavor.NewInsertBuilder().InsertInto("alr_meta")
+	updateAlrMeta.Cols("aco", "timestp").Values(aco, timestamp)
+	query, args := updateAlrMeta.Build()
+	query = fmt.Sprintf("%s RETURNING id", query)
+
+	var metaKey int
+	if err := tx.QueryRowEx(ctx, query, nil, args...).Scan(&metaKey); err != nil {
+		return err
+	}
+
+	// Update the alr table
+	cfs := &alrCopyFromSource{
+		ctx:       ctx,
+		metaKey:   metaKey,
+		timestamp: timestamp,
+		alrCount:  0,
+		alrRowLen: len(alrs),
+		alrs:      alrs,
+	}
+
+	fields := []string{"metakey", "mbi", "hic", "firstname", "lastname", "sex", "dob", "dod", "keyvalue"}
+	_, err = tx.CopyFrom(pgx.Identifier([]string{"alr"}), fields, cfs)
+	// If the copyfrom fails, attempt to rollback changes
+	if err != nil {
+		if err := tx.Rollback(); err != nil {
+			logrus.Warnf("Failed to rollback transaction %s", err.Error())
+		}
+		return err
+	}
+
+	// End the transaction
+	if err := tx.Commit(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *AlrRepository) GetAlr(ctx context.Context, ACO string, timestamp time.Time) ([]models.Alr, error) {
+
+	/* 	Query the alr_meta table to get foreign key */
+
+	tx, err := r.BeginTx(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Build the query
+	// sqlFlavor is from the repository.go file
+	meta := sqlFlavor.NewSelectBuilder()
+	meta.Select("id").From("alr_meta")
+	meta.Where(
+		meta.Equal("aco", ACO),
+		meta.GreaterEqualThan("timestp", timestamp),
+	)
+	query, args := meta.Build()
+
+	rows, err := tx.QueryContext(ctx, query, args...)
+	if err != nil {
+		// No rollback needed since nothing is begin added to DB
+		return nil, err
+	}
+
+	// Get the foreign keys from DB
+	var foreignKeys []interface{}
+	for rows.Next() {
+		var id int64
+		if err := rows.Scan(&id); err != nil {
+			return nil, err
+		}
+		foreignKeys = append(foreignKeys, id)
+	}
+
+	if len(foreignKeys) == 0 {
+		return nil, errors.New("No records were found")
+	}
+
+	/* 	End of alr_meta */
+
+	/* Query the alr table for ALR data */
+
+	// Build the query
+	alr := sqlFlavor.NewSelectBuilder()
+	alr.Select("id", "metakey", "mbi", "hic", "firstname", "lastname", "sex",
+		"dob", "dod", "keyvalue").From("alr")
+	alr.Where(
+		alr.In("metakey", foreignKeys...),
+	)
+	// Re-using variables from before
+	query, args = alr.Build()
+	// Using the foreign key, query the alr table
+	rows, err = tx.QueryContext(ctx, query, args...)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get ALR data
+	var alrs []models.Alr
+	for rows.Next() {
+		var alr models.Alr
+		var keyValueBytes []byte
+		if err := rows.Scan(&alr.ID, &alr.MetaKey, &alr.BeneMBI, &alr.BeneHIC,
```

Comment:
I know we had removed the timestamp field from the Alr struct, but I'm thinking it'll be important to have that information when we try to build the FHIR resource since it gives us the ability to present a timestamp on when the data was generated.

What're your thoughts on adding it the Alr model but not storing it directly in the database?